mapper/pojo-standalone/src/main/java/org/hibernate/search/mapper/pojo/standalone/loading/MassIdentifierLoader.java

```diff
@@ -4,10 +4,19 @@
  */
 package org.hibernate.search.mapper.pojo.standalone.loading;
 
+import java.util.OptionalLong;
+
 import org.hibernate.search.util.common.annotation.Incubating;
 
 /**
  * A loader for mass loading of entity identifiers, used in particular during mass indexing.
+ * <p>
+ * If the total count of identifiers to load is unknown, {@link #totalCount()} should remain unimplemented,
+ * i.e. return an empty optional. This will let the mass indexer know to increment the total on each new batch loaded instead
+ * of relying on the total being provided before the indexing starts.
```

Comment:
Hmm, to be honest ... IDK 🤷🏻‍♂️ 

Here's what I was thinking... There can be a mix, where we know total for one entity and not for the other.
If we index single type at a time, then not reporting the total works nice, we can say type1: remaining unknown; once indexed, we switch to the next type, and for it we'd report `Type2: Mass indexing is going to index N entities`
But what if we start indexing multiple types in parallel e.g. `.typesToIndexInParallel( 2 )`? Then we end up having `total = N + unknown` ... so in such scenario we wouldn't be able to report the remaining at all, which may be ok, IDK 😃 
That's why I thought if we increase the count before we feed the identifier to the queue it would look as if we started indexing a new type, and the count bumped.... 

On the other hand, something like this:
```
Indexed 99% 9900/10000 (remaining: 100)
Indexed 99% 10000/10100 (remaining: 100)
Indexed 99% 10100/10200 (remaining: 100)
Indexed 99% 10200/10300 (remaining: 100)
Indexed 99% 10300/10400 (remaining: 100)
```
is also confusing... 
but then it also may happen if we have a lot of types with small amount of documents and we index all, then the count will constantly increase and we'll see something similar in the logs ...

Reply:
One other thought I had was to add a new method to the monitor to report the number of loaded identifiers, but then discarded it, as it won't really solve anything for us here, and we already have the `entitiesLoaded(..)`/`documentsBuilt(..)` ...

Reply:
I think that's a good idea but:

1. I would just call it `typesToLoad` or `typeGroupsToLoad` or something like this, and call it at the very beginning. There isn't anything specific to unknown total, you're just giving information to the monitor about the number of type groups (i.e. the number of required calls to `addToTotalCount`).
2. You could even imagine overhauling the mass indexing monitor to make it more aware of type groups. E.g. `MassIndexingTypeGroupMonitor typeGroup(MassIndexingMonitorTypeGroupContext context)` and in Hibernate Search do `typeGroupMonitor = monitor.typeGroup(ctx);` and then use the `typeGroupMonitor` to report (type-group-scoped) information.
   Maybe this should be a separate Jira issue though.
3. I would display the number of remaining types in the logs.

So output would be as follows.

When we don't know the total at all:

```
Indexed 9900 (remaining: unknown)
Indexed 10000 (remaining: unknown)
Indexed 10100 (remaining: unknown)
Indexed 10200 (remaining: unknown)
Indexed 10300 (remaining: unknown)
```

When we know the total for the current group, but there are more groups:

```
Indexed 95% 9500/10000 (remaining: 500; 3 pending types)
Indexed 96% 9600/10000 (remaining: 400; 3 pending types)
Indexed 97% 9700/10000 (remaining: 300; 3 pending types)
Indexed 98% 9800/10000 (remaining: 200; 3 pending types)
Indexed 99% 9900/10000 (remaining: 100; 3 pending types)
```


When we know the total for the current group, and there are no more groups:

```
Indexed 95% 9500/10000 (remaining: 500)
Indexed 96% 9600/10000 (remaining: 400)
Indexed 97% 9700/10000 (remaining: 300)
Indexed 98% 9800/10000 (remaining: 200)
Indexed 99% 9900/10000 (remaining: 100)
```

--- 

BTW unrelated, to this issue, but if you're breaking the logging format, I'd recommend merging it all in one line. The current output is split over two lines, which can get quite annoying to read, and even just straight up confusing when indexing is fast and lines start interleaving.

The approximate remaining type could be useful as well, and is closely related to the "remaining count". I'd use remaining count divided by speed since last message, using a time unit that makes sense (`Duration` may help; surely there's a way to format that nicely without nonsensical precision?)

So maybe something like this?

```
Indexed: 9900. Per second: 142.2 instant, 123.5 overall. Remaining: unknown.
```

```
Indexed: 95% 9500/10000. Per second: 142.2 instant, 123.5 overall. Remaining: 500, 3 pending types. 
```

```
Indexed 95% 9500/10000. Per second: 142.2 instant, 123.5 overall. Remaining: 500, approximately 3.5 seconds.
```